django/db/models/functions/base.py

```diff
@@ -25,6 +25,10 @@ def __init__(self, expression, output_field):
     def as_sql(self, compiler, connection, **extra_context):
         if 'db_type' not in extra_context:
             extra_context['db_type'] = self._output_field.db_type(connection)
+        # Support CharField casting without max_length.
+        if (isinstance(self._output_field, fields.CharField) and
```

Comment:
This might be more extensible as a method on `CharField` rather than having field-specific checks in expressions.